drivers/input/input_cap12xx.c

```diff
@@ -0,0 +1,232 @@
+
+	return i2c_reg_write_byte_dt(i2c, REG_INTERRUPT_ENABLE, intr);
+}
+
+static int cap12xx_process(const struct device *dev)
+{
+	const struct cap12xx_config *config = dev->config;
+	struct cap12xx_data *data = dev->data;
+	int r;
+	uint8_t input;
+	uint8_t single_input_state;
+
+	/*
+	 * Clear INT bit to clear SENSOR INPUT STATUS bits.
+	 * Note that this is also required in polling mode.
+	 */
+	r = cap12xx_clear_interrupt(&config->i2c);
+	if (r < 0) {
+		return r;
+	}
+
+	r = i2c_reg_read_byte_dt(&config->i2c, REG_INPUT_STATUS, &input);
+	if (r < 0) {
+		return r;
+	}
+
+	for (uint8_t i = 0; i < config->input_channels; i++) {
+		single_input_state = input & BIT(i);
+		if (single_input_state != (data->prev_input_state & BIT(i))) {
+			input_report_key(dev, config->input_codes[i], single_input_state, true,
+					 K_FOREVER);
+		}
+	}
+	data->prev_input_state = input;
+
+	LOG_DBG("event: input: %d\n", input);
+
+	return 0;
+}
+
+static void cap12xx_work_handler(struct k_work *work)
+{
+	struct cap12xx_data *data = CONTAINER_OF(work, struct cap12xx_data, work);
+
+	cap12xx_process(data->dev);
+}
+
+static void cap12xx_isr_handler(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct cap12xx_data *data = CONTAINER_OF(cb, struct cap12xx_data, int_gpio_cb);
+
+	k_work_submit(&data->work);
+}
+
+#ifdef CONFIG_INPUT_CAP12XX_POLL
+static void cap12xx_timer_handler(struct k_timer *timer)
+{
+	struct cap12xx_data *data = CONTAINER_OF(timer, struct cap12xx_data, timer);
+
+	k_work_submit(&data->work);
+}
+#endif
+
+static int cap12xx_init(const struct device *dev)
+{
+	const struct cap12xx_config *config = dev->config;
+	struct cap12xx_data *data = dev->data;
+	int r;
+
+	if (!device_is_ready(config->i2c.bus)) {
+		LOG_ERR("I2C controller device not ready");
+		return -ENODEV;
+	}
+
+	data->dev = dev;
+
+	k_work_init(&data->work, cap12xx_work_handler);
+
+	if (config->int_gpio.port != NULL) {
+		if (!gpio_is_ready_dt(&config->int_gpio)) {
+			LOG_ERR("Interrupt GPIO controller device not ready");
+			return -ENODEV;
+		}
+
+		r = gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT);
+		if (r < 0) {
+			LOG_ERR("Could not confighure interrupt GPIO pin");
+			return r;
+		}
+
+		r = gpio_pin_interrupt_configure_dt(&config->int_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+		if (r < 0) {
+			LOG_ERR("Could not configure interrupt GPIO interrupt");
+			return r;
+		}
+
+		gpio_init_callback(&data->int_gpio_cb, cap12xx_isr_handler,
+				   BIT(config->int_gpio.pin));
+
+		r = gpio_add_callback(config->int_gpio.port, &data->int_gpio_cb);
+		if (r < 0) {
+			LOG_ERR("Could not set gpio callback");
+			return r;
+		}
+
+		r = cap12xx_clear_interrupt(&config->i2c);
+		if (r < 0) {
+			LOG_ERR("Could not clear interrupt");
+			return r;
+		}
+
+		r = cap12xx_enable_interrupt(&config->i2c, true);
+		if (r < 0) {
+			LOG_ERR("Could not configure interrupt");
+			return r;
+		}
+	}
+#ifdef CONFIG_INPUT_CAP12XX_POLL
+	else {
```

Comment:
confused, what if you don't specify the interrupt gpio and don't enable `CONFIG_INPUT_CAP12XX_POLL` either? Nothing would work right?

Either way bring the `ifdef` in the block, not a fan of splitting the `} else {` like that

Reply:
If you don't use polling, it's interrupts. If you don't specify an interrupt gpio, you get an error message.
But it will indeed not work.
I can change the code structure, this is based on the previous driver structure.

Reply:
> If you don't specify an interrupt gpio, you get an error message

Am I missing something? If `config->int_gpio.port` is NULL and `CONFIG_INPUT_CAP12XX_POLL=n` this would fall through and initialize with no way to work, right?

Reply:
No, you're right. If you do nothing at all, it will just add driver code. That's what the old driver did too.
What is the preferred behavior?
Should I LOG_ERR at run time?
There probably is a way to check if there is a device tree node at compile time too, you happen have an example of this?
Maybe I'll just #ifdef / #elseif the poll vs int parts depending on CONFIG_INPUT_CAP12XX_POLL.
Let me rewrite some more of that driver. 
(My intent was to just do a minimal change :-) )

Reply:
>  (My intent was to just do a minimal change :-) )

I  know sorry, told you you should have had a separate rename commit :-)

Two options:
- bring the ifdef into the else statement and if it's not set return an error
```c
} else {
#if CONFIG...POLL
        /* setup poll */
#else
        LOG_ERR("CONFIG_POLL must be enabled etc etc");
        return -EINVAL;
#endif